### ISSUE TITLE
Add ability to cache the downloaded web tiles

### ISSUE THREAD

**Issue by jorisvandenbossche** [darribas/contextily#46]
Since downloading the web tiles takes a bit of time, and typically you generate a figure multiple times while iterating on it, so I think that it would be nice to provide some caching mechanism for the downloaded web tiles.  

(+ other possible advantages: easier to use in case of flaky / limited internet connection, potentially more responsive if we could trigger matplotlib to re-draw when resizing)

I explored some options here: http://nbviewer.jupyter.org/gist/jorisvandenbossche/eef4dc7d7770e11e3056e8adb9d08b1c, and it improves plotting the `nybb` dataframe from 1s to 0.25 s.

The options I explored there:

- With https://github.com/reclosedev/requests-cache: since we already use `requests`, this might be an easy option.
- With https://github.com/joblib/joblib/ (what is eg used by sklearn to cache intermediate results; and also for parallelism, but that is less of our concern)

Thoughts? Other ideas to do this, or preferences?

**Comment by darribas** [comment]
Very interesting, thanks very much @jorisvandenbossche! I'd be positive to this in principle. A couple of ideas to discuss before we commit:

1. If all the machinery relies on `requests` I think we should be fine in terms of maintainability, right?
2. Where does the cache live and how is it managed in terms of reusability and of disposability once the session expires? I could see how some people might want to keep the cache indefinitely, but I'd also be wary that such an option would keep taking space on users' drives without them necessarily knowing it. Less commonly, but potentially relevant, since all of these sources rely on OSM, which is continuously improved, how'd the user know they're accessing the latest version?

**Comment by wpietri** [comment]
Hi! I recently had the same problem. My suggestion for the points in 2:

* Put the cache in /tmp/contextily/cache (using tempfile.gettempdir() to find the right platform-specific root)
* default to either no expiration or a small number of days (1-7)
* have a configuration option that lets people specify other location and cache policy

In case others have this problem and would like a temporary fix, for me it was sufficient to install squid and do `export HTTP_PROXY=127.0.0.1:3128` to speed things up dramatically (and be nicer to the place I'm getting my tiles).

**Comment by darribas** [comment]
Thanks very much for chipping in @wpietri. Just to leave it there for now, but something else we could add (if this ends up going in), is a simple `contextily` method that removes the cache so the user can always do something like `contextily.clean_cache()` and get rid of temporary files.

**Comment by jorisvandenbossche** [comment]
> If all the machinery relies on requests I think we should be fine in terms of maintainability, right?

Both options I mentioned above (requests-cache and joblib) are I think good external libraries, which should not give us much maintenance issues.

> Where does the cache live and how is it managed in terms of reusability and of disposability once the session expires? 

Yeah, this are certainly things we can play with. As @wpietri also mentioned, I think it comes down to choosing a good default, and then giving the option to change this. 
Some functionality to set the cache location and expiry policy (or a function to clear the cache as you mentioned) would be good.

**Comment by jorisvandenbossche** [comment]
> Both options I mentioned above (requests-cache and joblib) are I think good external libraries,

Although looking now again at requests-cache, it seems it has not seen any commit or release since 2 years, so maybe not that actively maintained. 
There is also `cachecontrol` (https://github.com/ionrock/cachecontrol), which has seen a bit more github activity lately, and maybe more importantly, is what is mentioned on the requests docs:http://docs.python-requests.org/en/master/community/recommended/?highlight=cache#cachecontrol

I quickly tested this `cachecontrol` as well, and it easily gives the same result as the other 2 I tested before. 
As far as I understand, this package should actually also use the header of the http get responce of the tile provider to determine if it can be cached or how long it can be cached (in case that is provided in the response)

**Comment by szmoore** [comment]
Sorry everyone, I did not see that there was already an issue and PR #57 for this before I did #68.

**Comment by jorisvandenbossche** [comment]
@szmoore no problem! We just need to discuss a bit and decide which approach we would want to follow.

**Comment by jorisvandenbossche** [comment]
For some context, there were a bunch of alternative open PRs:

- https://github.com/darribas/contextily/pull/57: using requests / CacheControl. Advantage: directly integrates with requests, which we are already using. Disadvantage: additional dependency on msgpack.
- https://github.com/darribas/contextily/pull/68 (@szmoore): manually creating temporary directory, and saving the images there. Advantage: no additional dependency, users can specify a directory to cache across sessions. 
- https://github.com/darribas/contextily/pull/79 (@uvchik): similar as the above
- https://github.com/darribas/contextily/pull/85: using joblib. Advantage: only small code change to have basic caching. Additional dependency but an "easy" one (pure python + already used by other popular packages such as scikit-learn). Disadvantage: does not provide easy option to keep cache accross sessions or specify custom cache directory (also does not use human readable file names).

We just merged #85 (using joblib) to at least have basic caching in the upcoming release. 

However, as mentioned above, it doesn't suite all needs. For example, the use case of specifying your custom directory with cached files (or manually downloaded files) does not seem to be easy. Also setting the option of caching across sessions is not implemented (this probably needs to ability to specify a custom directory as mentioned above). 
So longer term, we might still be interested in one of the other approaches, if those use cases are deemed important (in any case, the most urgent thing, basic caching to improve performance within one session, is covered). In such a case, I think we should start with specifying the use cases and interface we want to provide to users, before delving in the code.

**Comment by uvchik** [comment]
The need of a manual download came up, because I was not able to download "https://a.tile.openstreetmap.org/tileZ/tileX/tileY.png" via contextily, but I was able to download it using my browser. Actuially I think this should be solved in another way. Manual download was just a work around.

So readable filenames is not that important, even though I like it.

But I definitely need caching across session because, I do need an easy possibility to work offline.

**Comment by darribas** [comment]
Thanks @uvchik . We also recently moved to the standard way of parsing tile addresses (with the `{X}` instead of `tileX`, etc.). Could that be why you couldn't download it via contextily?

**Comment by uvchik** [comment]
> Thanks @uvchik . We also recently moved to the standard way of parsing tile addresses (with the `{X}` instead of `tileX`, etc.). Could that be why you couldn't download it via contextily?

I don't think so, because the resulting URL was fine in both ways.
e.g. https://a.tile.openstreetmap.org/17/68852/41965.png

It worked for other tile servers such as http://a.tile.openstreetmap.fr/osmfr/ but it failed for  https://a.tile.openstreetmap.org/. It is not a big problem and I may open an extra issue on that.

the ability to work offline is more important for me.

**Comment by darribas** [comment]
Thanks for the feedback.

> The ability to work offline is still more important for me.

That's great feedback. The current approach we take to this is to save a map (merged tiles) as a raster file (e.g. with `bounds2raster`) and use that file later on as the source. That way we can support both use of arbitrary rasters and webmaps saved as such with pretty much the same machinery, making it easier for maintainability.

Just out of curiosity, why'd you need the tiles locally instead of the merged `.tif`?

**Comment by uvchik** [comment]
> Just out of curiosity, why'd you need the tiles locally instead of the merged `.tif`?

Using the tiles is more flexible and I really like that I do not have to change the structure. Just one parameter decides whether I use the cached files or the online files. I can move my plot a bit to the east and it will work directly (the function may download some additional tiles).

**Comment by darribas** [comment]
I know it's a different approach, but something I've been thinking of getting in is #86 which would allow you to download into a single raster (or have a very large raster file locally from other sources, like the night lights from NASA) a larger extent and use `add_basemap` to only read the extent you require efficiently. Again, it's a different approach but might do a similar outcome?

**Comment by uvchik** [comment]
You are right, there are different way to reach the same goal. I just think that my approach is very easy from the user perspective. If you have a running function with tiles from the server you just have to add a path to download all tiles and use them offline the next run. You do not have to change the parameters or to use a different function. It is one setup for both use cases.

**Comment by darribas** [comment]
OK, I think I start seeing your use case, it's more a case when you're trying to set almost a local server to interact with, correct?

**Comment by uvchik** [comment]
Yes, you could see it this way.

**Comment by jorisvandenbossche** [comment]
> The need of a manual download came up, because I was not able to download "https://a.tile.openstreetmap.org/tileZ/tileX/tileY.png" via contextily, but I was able to download it using my browser. Actuially I think this should be solved in another way. Manual download was just a workaround.

Do you still remember in which case you ran into this? As that sounds like a potential bug if the actual url works fine.

> I do not know how to add permanent caching to the joblib approach, so I cannot help there but I am not necessarily attached to my approach but would help if I can.

It's maybe a bit hacking into joblib, but it seems this might be possible in a different way as I tried in the PR (instead of overwriting the full Memory object, only overwriting the location of the existing memory object):

```
In [37]: import contextily    

In [38]: contextily.tile.memory                                                                                                                                                                                    
Out[38]: Memory(location=/tmp/tmptdthyfsc/joblib)

In [40]: contextily.tile.memory.store_backend.location
Out[40]: '/tmp/tmptdthyfsc/joblib'

In [41]: contextily.tile.memory.store_backend.location = './joblib'   
```

and then running some download queries with contextily seems to correctly use that specified directory as the cache (we should maybe ask the joblib devs if that is somewhat safe to do).

We could provide a wrapper function (with some safety checks) like `contextily.set_cache_dir` that basically does that.

**Comment by uvchik** [comment]
Thanks I will give it a try. Unfortunately I will not have time the next two weeks but will go on afterwards.

**Comment by uvchik** [comment]
@jorisvandenbossche I tested your approach and everything works fine:

With this solution we rely on joblib but as long as this project is active this will not be a problem.
Furthermore the results are not human readable but as the problems with the default url of OSM no longer occur there is no need for human readable tiles.

Thanks for implementing this feature. I will close my PR (#79) because it is obsolete now.

**Comment by uvchik** [comment]
Shall we announce this solution somewhere in the contextily or geopandas documentation?

For me this is very valuable because it speeds up plotting
As there have been different PRs with this aim, it seems to be a needed feature not only for me.

**Comment by jorisvandenbossche** [comment]
I would personally still wrap it in a function that contextily exposes (eg `set_cache_dir`), before officially documenting it. 
That way we can hide somewhat the joblib internals (or also deal with that if we would change the underlying method in the future while keeping the same user API).

**Comment by uvchik** [comment]
Good idea.

1. I would name it `use_cache` because one has to execute it every time one wants to plot. For me 'set' indicates, that I have to execute it once. But I am not a native speaker and I am open to other names.

```python
def use_cache(path):
    """
    Parameters
    ----------
    path : str
        Path to the cache directory.
    """
    tile.memory.store_backend.location = path
```

2. I am not sure whether we should add a default value. We could add something like `os.path.join(os.path.expanduser('~'), '.contextily')`, but it might be clearer to have no default value.


3. I would like to add this function in the `add_basemap` function. Therefore, I would like to add a `cache_path` parameter there. The reason for that is that I think that caching should be the default behaviour but at least very obvious to the user.

**Comment by abubelinha** [comment]
Hi.  I think this is the default behaviour now.  
May I ask how long cached tiles last?  Is this configurable somehow?

I would like to know how long a script would keep working if I go offline for vacations with my laptop and try to regenerate maps of a previously mapped area (same zoom levels).

Thanks
@abubelinha

**Comment by martinfleis** [comment]
We're using ´joblib.Memory` to deal with the cache and I believe the cache is valid only within a running script/notebook. So it helps if you're having multiple calls to contextily plotting the same area but it does not persist the data between sessions. So I believe that your use case of mapping offline would not work. But you can fetch the tiles and work with them from a disk. See https://contextily.readthedocs.io/en/latest/working_with_local_files.html

**Comment by abubelinha** [comment]
Thanks. So using `bound2raster` seems to be the trick.
I am not sure I understand the example in docs. I think there is a mistake in it:

> Or download a basemap to a local file and then plot it:
> 
> ```Python
> source = 'virginia.tiff'
> _ = cx.bounds2raster(*db.total_bounds, zoom=6, source=source) # source=source ?????
> ax = db.plot(alpha=0.5, color='k', figsize=(6, 6))
> cx.add_basemap(ax, source=source)
> plt.show()
> ```

Shouldn't we first save tiles from provider url to a local geotiff path, before we can read tiles from it??
```Python
cx.bounds2raster(source=provider_url, path='virginia.tiff')
cx.add_basemap(ax, source='virginia.tiff')
```

**Comment by martinfleis** [comment]
> Shouldn't we first save tiles from provider url to a local geotiff path, before we can read tiles from it??

Yes. The example assumes you want to use different data as tiles, not stuff saved from web. Also, make sure downloading in this form does not violate the terms and conditions of a selected provider.

**Comment by abubelinha** [comment]
Thanks.
Regarding terms and conditions I was also thinking about rate download limits.  
Do you somehow consider this in providers definition? 

I don't know if contextily takes care of delaying tile requests to avoid overloading servers, or there is some option for me to do it depending on which tile servers I am using.
